hel_django_auditlog_extra/mixins.py

```diff
@@ -0,0 +1,113 @@
+
+        The changelist in the response's context_data (i.e. "cl") should be set by the
+        super().changelist_view call, see e.g.
+        https://github.com/django/django/blob/4.2.16/django/contrib/admin/options.py#L2071
+
+        Args:
+            request: The HTTP request object.
+            extra_context: Optional extra context to pass to the template.
+
+        Returns:
+            The response from the superclass's `changelist_view` method.
+        """
+        response = super().changelist_view(request, extra_context)
+        if self.write_accessed_from_list_view:
+            changelist = response.context_data["cl"]
+            for obj in changelist.result_list:
+                accessed.send(sender=obj.__class__, instance=obj, actor=request.user)
+        return response
+
+    def change_view(self, request, object_id, form_url="", extra_context=None):
+        """
+        Logs the access event when the change view of an object is accessed.
+
+        Args:
+            request: The HTTP request object.
+            object_id: The ID of the object being changed.
+            form_url: The URL of the change form.
+            extra_context: Optional extra context to pass to the template.
+
+        Returns:
+            The response from the superclass's `change_view` method.
+        """
+        obj = self.get_object(request, object_id)
+        if obj is not None:
+            accessed.send(sender=obj.__class__, instance=obj, actor=request.user)
+        return super().change_view(request, object_id, form_url, extra_context)
+
+    def history_view(self, request, object_id, extra_context=None):
+        """
+        Logs the access event when the history view of an object is accessed.
+
+        Args:
+            request: The HTTP request object.
+            object_id: The ID of the object whose history is being viewed.
+            extra_context: Optional extra context to pass to the template.
+
+        Returns:
+            The response from the superclass's `history_view` method.
+        """
+        obj = self.get_object(request, object_id)
+        if obj is not None:
+            accessed.send(sender=obj.__class__, instance=obj, actor=request.user)
+        return super().history_view(request, object_id, extra_context)
+
+    def delete_view(self, request, object_id, extra_context=None):
+        """
+        Logs the access event when the delete view of an object is accessed.
+
+        Args:
+            request: The HTTP request object.
+            object_id: The ID of the object being deleted.
+            extra_context: Optional extra context to pass to the template.
+
+        Returns:
+            The response from the superclass's `delete_view` method.
+        """
+        obj = self.get_object(request, object_id)
+        if obj is not None:
+            accessed.send(sender=obj.__class__, instance=obj, actor=request.user)
+        return super().delete_view(request, object_id, extra_context)
```

Comment:
Could all this be replaced with `get_object` overriding?

```suggestion
    def get_object(self, request, object_id, from_field=None):
        obj = super().get_object(request, object_id, from_field)
        if obj is not None:
            accessed.send(sender=obj.__class__, instance=obj, actor=request.user)
        return obj
```

Reply:
the get_object would be more reusable.

Reply:
This is now done in https://github.com/City-of-Helsinki/kukkuu/pull/440/commits/0597963ef346802db541eca894e9f58778130e9e. There was a catch in this get_object implementation: it can be called many times, so a contextvar or request object manipulation was needed. I'm now suggesting contextvar to prevent duplicate signals from that action.